### PR TITLE
Support vim modelines

### LINKED ISSUE #886: Support vim modelines
It would be great if VsVim handled vim modelines.  I often use modelines to set up per-file tab/indent preferences. For example:

```
# vim:ts=4:sw=4:et
```

Thanks for this excellent extension!

### LINKED ISSUE #2548: Vim help files use the user's tab settings rather than those specified in the files
Inline vim help are classified as Visual Studio text files and currently display using the user's tab settings, which are often some other value than eight. The help files use tabs extensively and expect the setting `:set tabstop=8` to be in effect. This is accomplished via the modeline mechanism, such as this line near the bottom of help.txt:

```
 vim:tw=78:isk=!-~,^*,^\|,^\":ts=8:noet:ft=help:norl:
```

Although full modeline support is desirable, what is mostly needed for this issue is automatically setting `ts=8` in vim help files.

### PR DESCRIPTION
**Pull request by ricksladkey**
#### Issues

- Closes #886
- Closes #2548

### CODE REVIEW

File: Src/VimCore/ModeLineInterpreter.fs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+               result
+
+        // Try to process the first few and last few lines as modelines.
+        let tryProcessModeLines modeLines =
+            let lineCount = _textBuffer.CurrentSnapshot.LineCount
+            let snapshot = _textBuffer.CurrentSnapshot
+            seq {
+                yield seq { 0 .. min (modeLines - 1) (lineCount - 1) }
+                yield seq { max modeLines (lineCount - modeLines) .. lineCount - 1 }
+            }
```
Comment (jaredpar) [inline Src/VimCore/ModeLineInterpreter.fs]: Isn't this seq just equivalent to all the line numbers in the snapshot?

File: Src/VimCore/ModeLineInterpreter.fs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            else
+                false
+
+        // Split the options string into fields.
+        let splitFields (options: string) =
+            options.Replace(@"\:", ":").Split(' ', '\t')
+
+        // Process the "first" format of modeline, i.e. "vim: ...".
+        let processFirstForm modeLine =
+            let m = Regex.Match(modeLine, _firstPattern)
```
Comment (jaredpar) [inline Src/VimCore/ModeLineInterpreter.fs]: This means that for every line we're constructing the `Regex` object for `_firstPattern` and then executing it fresh against the current line. Worried that could be expensive. Did you consider making the fields be of type `Regex` and using the compiled option to ensure speedy execution here?

File: Src/VimCore/ModeLineInterpreter.fs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                    |> Seq.map (fun m -> splitFields m.Groups.[1].Value)
+                    |> Seq.concat
+                    |> Seq.tryFind (fun option -> not (processOption option))
+                Some modeLine, firstBadOption
+            else
+                None, None
+
+        // Process the "second" format of modeline, i.e. "vim: set ... :".
+        let processSecondForm modeLine =
+            let m = Regex.Match(modeLine, _secondPattern)
```
Comment (jaredpar) [inline Src/VimCore/ModeLineInterpreter.fs]: Same feedback about constructing `Regex` objects.

File: Src/VimCore/ModeLineInterpreter.fs (hunk 1 above)
Comment (ricksladkey) [reply Src/VimCore/ModeLineInterpreter.fs]: The default for `modelines` is `5`, so typically we search the first five and the last five lines. The min / max calculations here handle files that are shorter than 10 lines long.  Let's say there are 100 lines. Then the expression is like:

```
seq {
    yield seq { 0 .. 4 }
    yield seq { 95 .. 99 }
}
```

File: Src/VimCore/ModeLineInterpreter.fs (hunk 2 above)
Comment (ricksladkey) [reply Src/VimCore/ModeLineInterpreter.fs]: That's a good point; I'll make that change.

File: Src/VimCore/ModeLineInterpreter.fs (hunk 1 above)
Comment (ricksladkey) [reply Src/VimCore/ModeLineInterpreter.fs]: Added a test to prove that a modeline in the middle of a large file is not processed.

File: Src/VimCore/ModeLineInterpreter.fs (hunk 3 above)
Comment (ricksladkey) [reply Src/VimCore/ModeLineInterpreter.fs]: Changed these all to static compiled regular expression objects.

**Pull request merged**